docs/release_guide.md

```diff
@@ -54,13 +54,13 @@ If there are other commits on the `main` branch to be included in the release (s
 
 Create a new commit on the patch branch to update the hardcoded version information in the following locations:
 
-* The container image tag in [charts/osm/values.yaml](/charts/osm/values.yaml)
+* The container image digests `OpenServiceMesh.image.digest` for images in [charts/osm/values.yaml](/charts/osm/values.yaml)
```

Comment:
@nojnhuh That's a valid question I didn't address in my commit message, but I think the release workflow should be updated to:
1. Publish the images with the appropriate release tag
2. Make the changes to retrieve the digest from the published images as a part of building the binaries for the release

Is that something that is feasible?

Reply:
> 1. Publish the images with the appropriate release tag

So we move image builds from the CI environment to local workstations? I don't see that being a major problem, but does open up opportunities to make mistakes.

> 2. Make the changes to retrieve the digest from the published images as a part of building the binaries for the release

Then where do you verify the retrieved digests are the same as the digests of the images you built?

> Is that something that is feasible?

Probably? I'm not totally sure. I'd like to see what it would take to get fully reproducible builds, since if we could leverage that then it wouldn't take as much tweaking to the overall process.

Reply:
> > 1. Publish the images with the appropriate release tag
> 
> So we move image builds from the CI environment to local workstations? I don't see that being a major problem, but does open up opportunities to make mistakes.
> 
No, we could make it a part of the CI but it could be a pre-release step, where we make a code change to `constants.go` to update the image references and publish the images with the appropriate tags. 

> > 1. Make the changes to retrieve the digest from the published images as a part of building the binaries for the release
> 
> Then where do you verify the retrieved digests are the same as the digests of the images you built?
> 
This can be retrieved from the image manifest. We wouldn't need to verify anything else other than the fact that the digest matches the image that was published. This is much like updating the tags currently, with correctness of the update verified through code reviews.

> > Is that something that is feasible?
> 
> Probably? I'm not totally sure. I'd like to see what it would take to get fully reproducible builds, since if we could leverage that then it wouldn't take as much tweaking to the overall process.

What's the use case for reproducible builds? I am not sure how using image digests could play a role here. Could you share an example?

Reply:
> No, we could make it a part of the CI but it could be a pre-release step, where we make a code change to `constants.go` to update the image references and publish the images with the appropriate tags.

There's still the nuance where the init image needs to be built before the injector image so the injector know the init image's digest. That's certainly possible to configure, just might be a bit convoluted. What would the trigger for this CI job be?

> This can be retrieved from the image manifest. We wouldn't need to verify anything else other than the fact that the digest matches the image that was published. This is much like updating the tags currently, with correctness of the update verified through code reviews.

I'm not exactly sure which digests you're suggesting to compare. I was trying to say we should make sure the digests we put in values.yaml and elsewhere come straight from the images built in CI and not after some push and pull if we're trying to protect against images that have been modified in Docker Hub.

> What's the use case for reproducible builds? I am not sure how using image digests could play a role here. Could you share an example?

If builds are reproducible, then we can build the images locally, see what the digests are, update image references accordingly like we do today with tags, then tag and push that commit. The CI would then build and publish images that result in the same digests since it's working off of the same tagged commit.

Reply:
@nojnhuh the digest is calculated on push to a registry. So we will not be able to calculate this beforehand. The goal is to ensure the digest is correct and bake that into the chart values **before** cutting a release.

Reply:
> @nojnhuh the digest is calculated on push to a registry. So we will not be able to calculate this beforehand.

The digest is a property of the image itself, so we only need to build it to know what the digest is. No push is necessary. `docker inspect <image> -f '{{ .ID }}'`

> The goal is to ensure the digest is correct and bake that into the chart values before cutting a release.

Yeah I think we're agreeing here. Let's make sure all the CI changes for release are captured in this PR. Commenting on code is much easier than English 😄